// jvm/lib/core/jvm/MethodInfo.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MethodInfo = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _get = function get(object, property, receiver) { if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { return get(parent, property, receiver); } } else if ("value" in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } };

exports.parameterParser = parameterParser;

var _lodash = require('lodash');

var _ = _interopRequireWildcard(_lodash);

var _AccessFlags = require('./AccessFlags');

var _MemberInfo2 = require('./MemberInfo');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/**
 * Parses the parameter types from a method description.
 * @param  {string} desc - method descriptor
 * @return {Array<string>}
 */
function parameterParser(desc) {
  var params = [];
  var paramString = desc.slice(desc.indexOf('(') + 1, desc.indexOf(')'));
  var arrayStr = '';
  for (var i = 0; i < paramString.length;) {
    var token = paramString[i];
    switch (token) {
      case '[':
        arrayStr += '[';
        i++;
        break;
      case 'B':
      case 'C':
      case 'D':
      case 'F':
      case 'I':
      case 'J':
      case 'S':
      case 'Z':
        params.push(arrayStr + token);
        arrayStr = '';
        i++;
        break;
      case 'L':
        var end = paramString.indexOf(';', i);
        var param = paramString.slice(i, end + 1);
        params.push(arrayStr + param);
        arrayStr = '';
        i += param.length;
        break;
    }
  }

  return params;
};

/**
 * Wrapper for JVM Class File method entry.
 */

var MethodInfo = function (_MemberInfo) {
  _inherits(MethodInfo, _MemberInfo);

  /**
   * 
   * @param  {ClassInfo} classInfo - The class the given method is in
   * @param  {Number} accessFlags - access flags bit string
   * @param  {string} name - name of this entry in the ClassFile
   * @param  {string} descriptor - type descriptor
   */
  function MethodInfo(classInfo, flags, name, descriptor) {
    _classCallCheck(this, MethodInfo);

    /** @type {ClassInfo} class this method is within */
    var _this = _possibleConstructorReturn(this, (MethodInfo.__proto__ || Object.getPrototypeOf(MethodInfo)).call(this, flags, name));

    _this._classInfo = classInfo;
    /** @type {string} type descriptor */
    _this._descriptor = descriptor;
    return _this;
  }

  /**
   * Access the class this member is within
   * @return {ClassInfo}
   */


  _createClass(MethodInfo, [{
    key: 'toObject',


    /**
     * Serialized version of this class without circular references.
     * @return {Object}
     */
    value: function toObject() {
      return _.merge(_get(MethodInfo.prototype.__proto__ || Object.getPrototypeOf(MethodInfo.prototype), 'toObject', this).call(this), {
        desc: this.desc,
        instructions: _.map(this.instructions, function (insn) {
          return insn.toObject();
        })
      });
    }
  }, {
    key: 'classInfo',
    get: function get() {
      return this._classInfo;
    }

    /**
     * JVM type descriptor
     * @return {string}
     */

  }, {
    key: 'desc',
    get: function get() {
      return this._descriptor;
    }

    /**
     * Checks if the method should not have a method body by seeing if the
     * Synthetic or Abstract access flags are set.
     * @return {Boolean}
     */

  }, {
    key: 'hasNoMethodBody',
    get: function get() {
      return this.hasFlag(_AccessFlags.ACC_SYNTHETIC) || this.hasFlag(_AccessFlags.ACC_ABSTRACT);
    }

    /**
     * Inverse of {@link MethodInfo.hasNoMethodBody}.
     * @return {Boolean}
     */

  }, {
    key: 'hasMethodBody',
    get: function get() {
      return !this.hasNoMethodBody;
    }

    /**
     * Getter for the collection of instructions belonging to a method.
     * @return {Array<AbstractInstruction>}
     */

  }, {
    key: 'instructions',
    get: function get() {
      var code = this.findAttributeByName('Code');
      if (code) {
        return code.decoded;
      }
      return [];
    }

    /**
     * Parses the method descriptor's parameters.
     * @return {Array<string>}
     */

  }, {
    key: 'parameters',
    get: function get() {
      return parameterParser(this.desc);
    }

    /**
     * Parses the return type from the method's descriptor.
     * @return {string}
     */

  }, {
    key: 'returnType',
    get: function get() {
      return this.desc.slice(this.desc.lastIndexOf(')') + 1);
    }
  }]);

  return MethodInfo;
}(_MemberInfo2.MemberInfo);

exports.MethodInfo = MethodInfo;
