// jvm/lib/core/jvm/instructions/ImmediateShortInstruction.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _get = function get(object, property, receiver) { if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { return get(parent, property, receiver); } } else if ("value" in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } };

var _AbstractInstruction2 = require('./AbstractInstruction');

var _AbstractInstruction3 = _interopRequireDefault(_AbstractInstruction2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; } /** @ignore */


var ImmediateShortInstruction = function (_AbstractInstruction) {
  _inherits(ImmediateShortInstruction, _AbstractInstruction);

  function ImmediateShortInstruction(methodInfo, idx, opcode) {
    _classCallCheck(this, ImmediateShortInstruction);

    var _this = _possibleConstructorReturn(this, (ImmediateShortInstruction.__proto__ || Object.getPrototypeOf(ImmediateShortInstruction)).call(this, methodInfo, idx, opcode));

    _this.val = 0;
    return _this;
  }

  _createClass(ImmediateShortInstruction, [{
    key: 'read',
    value: function read(buffer) {
      _get(ImmediateShortInstruction.prototype.__proto__ || Object.getPrototypeOf(ImmediateShortInstruction.prototype), 'read', this).call(this, buffer);
      this.val = buffer.short();
    }
  }, {
    key: 'write',
    value: function write(buffer) {
      _get(ImmediateShortInstruction.prototype.__proto__ || Object.getPrototypeOf(ImmediateShortInstruction.prototype), 'write', this).call(this, buffer);
      buffer.writeShort(this.val);
    }
  }, {
    key: 'toObject',
    value: function toObject() {
      return _get(ImmediateShortInstruction.prototype.__proto__ || Object.getPrototypeOf(ImmediateShortInstruction.prototype), 'toObject', this).call(this, {
        val: this.val,
        size: this.size
      });
    }
  }, {
    key: 'size',
    get: function get() {
      return _get(ImmediateShortInstruction.prototype.__proto__ || Object.getPrototypeOf(ImmediateShortInstruction.prototype), 'size', this) + 2;
    }
  }]);

  return ImmediateShortInstruction;
}(_AbstractInstruction3.default);

exports.default = ImmediateShortInstruction;


module.exports = ImmediateShortInstruction;
