// jvm/lib/core/jvm/instructions/InvokeInterfaceInstruction.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _get = function get(object, property, receiver) { if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { return get(parent, property, receiver); } } else if ("value" in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } };

var _ImmediateShortInstruction = require('./ImmediateShortInstruction');

var _ImmediateShortInstruction2 = _interopRequireDefault(_ImmediateShortInstruction);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; } /** @ignore */


var InvokeInterfaceInstruction = function (_ImmediateShortInstru) {
  _inherits(InvokeInterfaceInstruction, _ImmediateShortInstru);

  function InvokeInterfaceInstruction(methodInfo, idx, opcode) {
    _classCallCheck(this, InvokeInterfaceInstruction);

    var _this = _possibleConstructorReturn(this, (InvokeInterfaceInstruction.__proto__ || Object.getPrototypeOf(InvokeInterfaceInstruction)).call(this, methodInfo, idx, opcode));

    _this.count = 0;
    return _this;
  }

  _createClass(InvokeInterfaceInstruction, [{
    key: 'read',
    value: function read(buffer) {
      _get(InvokeInterfaceInstruction.prototype.__proto__ || Object.getPrototypeOf(InvokeInterfaceInstruction.prototype), 'read', this).call(this, buffer);
      this.count = buffer.byte();
      // the next byte is always zero and thus discarded.
      buffer.byte();
    }
  }, {
    key: 'write',
    value: function write(buffer) {
      _get(InvokeInterfaceInstruction.prototype.__proto__ || Object.getPrototypeOf(InvokeInterfaceInstruction.prototype), 'write', this).call(this, buffer);
      buffer.writeByte(this.count);
      buffer.writeByte(0);
    }
  }, {
    key: 'toObject',
    value: function toObject() {
      return _get(InvokeInterfaceInstruction.prototype.__proto__ || Object.getPrototypeOf(InvokeInterfaceInstruction.prototype), 'toObject', this).call(this, {
        count: this.count,
        size: this.size
      });
    }
  }, {
    key: 'toString',
    value: function toString() {
      return _get(InvokeInterfaceInstruction.prototype.__proto__ || Object.getPrototypeOf(InvokeInterfaceInstruction.prototype), 'toString', this).call(this, { count: this.count });
    }
  }, {
    key: 'size',
    get: function get() {
      return _get(InvokeInterfaceInstruction.prototype.__proto__ || Object.getPrototypeOf(InvokeInterfaceInstruction.prototype), 'size', this) + 2;
    }
  }]);

  return InvokeInterfaceInstruction;
}(_ImmediateShortInstruction2.default);

exports.default = InvokeInterfaceInstruction;


module.exports = InvokeInterfaceInstruction;
