// jvm/lib/core/jvm/instructions/LookupSwitchInstruction.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _get = function get(object, property, receiver) { if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { return get(parent, property, receiver); } } else if ("value" in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } };

var _lodash = require('lodash');

var _ = _interopRequireWildcard(_lodash);

var _PaddedInstruction2 = require('./PaddedInstruction');

var _PaddedInstruction3 = _interopRequireDefault(_PaddedInstruction2);

var _OffsetPair = require('../OffsetPair');

var _OffsetPair2 = _interopRequireDefault(_OffsetPair);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/** @ignore */

/** @ignore */


var LookupSwitchInstruction = function (_PaddedInstruction) {
  _inherits(LookupSwitchInstruction, _PaddedInstruction);

  function LookupSwitchInstruction(methodInfo, idx, opcode) {
    _classCallCheck(this, LookupSwitchInstruction);

    var _this = _possibleConstructorReturn(this, (LookupSwitchInstruction.__proto__ || Object.getPrototypeOf(LookupSwitchInstruction)).call(this, methodInfo, idx, opcode));

    _this.offsetPairs = [];
    _this.defaultOffset = 0;
    return _this;
  }

  _createClass(LookupSwitchInstruction, [{
    key: 'read',
    value: function read(buffer) {
      _get(LookupSwitchInstruction.prototype.__proto__ || Object.getPrototypeOf(LookupSwitchInstruction.prototype), 'read', this).call(this, buffer);
      while (this.offsetPairs.length > 0) {
        offsetPairs.pop();
      }
      this.defaultOffset = buffer.int();
      var pairCount = buffer.int();
      for (var i = 0; i < pairCount; i++) {
        var match = buffer.int();
        var offset = buffer.int();
        this.offsetPairs.push(new _OffsetPair2.default(match, offset));
      }
    }
  }, {
    key: 'write',
    value: function write(buffer) {
      _get(LookupSwitchInstruction.prototype.__proto__ || Object.getPrototypeOf(LookupSwitchInstruction.prototype), 'write', this).call(this, buffer);
      buffer.writeInt(this.defaultOffset);
      var pairCount = this.offsetPairs.length;
      buffer.writeInt(pairCount);
      _.each(this.offsetPairs, function (offsetPair) {
        buffer.writeInt(offsetPair.match);
        buffer.writeInt(offsetPair.offset);
      });
    }
  }, {
    key: 'toObject',
    value: function toObject() {
      return _get(LookupSwitchInstruction.prototype.__proto__ || Object.getPrototypeOf(LookupSwitchInstruction.prototype), 'toObject', this).call(this, {
        offset_pairs: this.offsetPairs,
        default_offset: this.defaultOffset,
        size: this.size
      });
    }
  }, {
    key: 'toString',
    value: function toString() {
      return _get(LookupSwitchInstruction.prototype.__proto__ || Object.getPrototypeOf(LookupSwitchInstruction.prototype), 'toString', this).call(this, { defaultOffset: this.defaultOffset, offsetPairs: this.offsetPairs });
    }
  }, {
    key: 'size',
    get: function get() {
      return _get(LookupSwitchInstruction.prototype.__proto__ || Object.getPrototypeOf(LookupSwitchInstruction.prototype), 'size', this) + 8 + 8 * this.offsetPairs.length;
    }
  }]);

  return LookupSwitchInstruction;
}(_PaddedInstruction3.default);

exports.default = LookupSwitchInstruction;


module.exports = LookupSwitchInstruction;
