// jvm/lib/core/jvm/instructions/TableSwitchInstruction.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _get = function get(object, property, receiver) { if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { return get(parent, property, receiver); } } else if ("value" in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } };

var _lodash = require('lodash');

var _ = _interopRequireWildcard(_lodash);

var _PaddedInstruction2 = require('./PaddedInstruction');

var _PaddedInstruction3 = _interopRequireDefault(_PaddedInstruction2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/** @ignore */


var TableSwitchInstruction = function (_PaddedInstruction) {
  _inherits(TableSwitchInstruction, _PaddedInstruction);

  function TableSwitchInstruction(methodInfo, idx, opcode) {
    _classCallCheck(this, TableSwitchInstruction);

    var _this = _possibleConstructorReturn(this, (TableSwitchInstruction.__proto__ || Object.getPrototypeOf(TableSwitchInstruction)).call(this, methodInfo, idx, opcode));

    _this.defaultOffset = 0;
    _this.lowByte = 0;
    _this.highByte = 0;
    _this.jumpOffsets = [];
    return _this;
  }

  _createClass(TableSwitchInstruction, [{
    key: 'read',
    value: function read(buffer) {
      _get(TableSwitchInstruction.prototype.__proto__ || Object.getPrototypeOf(TableSwitchInstruction.prototype), 'read', this).call(this, buffer);

      this.defaultOffset = buffer.int();
      this.lowByte = buffer.int();
      this.highByte = buffer.int();

      this.jumpOffsets = [];
      var offsetCount = this.highByte - this.lowByte + 1;
      for (var i = 0; i < offsetCount; i++) {
        this.jumpOffsets[i] = buffer.int();
      }
    }
  }, {
    key: 'write',
    value: function write(buffer) {
      _get(TableSwitchInstruction.prototype.__proto__ || Object.getPrototypeOf(TableSwitchInstruction.prototype), 'write', this).call(this, buffer);
      buffer.writeInt(this.defaultOffset);
      buffer.writeInt(this.lowByte);
      buffer.writeInt(this.highByte);
      _.each(this.jumpOffsets, function (jumpOffset) {
        buffer.writeInt(jumpOffset);
      });
    }
  }, {
    key: 'toObject',
    value: function toObject() {
      return _get(TableSwitchInstruction.prototype.__proto__ || Object.getPrototypeOf(TableSwitchInstruction.prototype), 'toObject', this).call(this, {
        default_offset: this.defaultOffset,
        low_byte: this.lowByte,
        high_byte: this.highByte,
        jump_offsets: this.jumpOffsets,
        size: this.size
      });
    }
  }, {
    key: 'toString',
    value: function toString() {
      return _get(TableSwitchInstruction.prototype.__proto__ || Object.getPrototypeOf(TableSwitchInstruction.prototype), 'toString', this).call(this, {
        defaultOffset: this.defaultOffset,
        jumpOffsets: this.jumpOffsets,
        lowByte: this.lowByte,
        highByte: this.highByte
      });
    }
  }, {
    key: 'size',
    get: function get() {
      return _get(TableSwitchInstruction.prototype.__proto__ || Object.getPrototypeOf(TableSwitchInstruction.prototype), 'size', this) + 12 + 4 * this.jumpOffsets.length;
    }
  }]);

  return TableSwitchInstruction;
}(_PaddedInstruction3.default);

exports.default = TableSwitchInstruction;


module.exports = TableSwitchInstruction;
